Trogdor Improvements: execute Trogdor tasks directly and run the Trogdor worker in a separate process

yangxi opened the pull request:
This PR adds two features: 1) directly execute Trogdor tasks without setting up coordinators and agents, 2) make agents be able to execute shell commands as the worker in another process, so we have the flexibility of running workloads, such as non-Java clients, docker images, etc.

* Directly execute Trogdor tasks (feature 1).
    * Add TrogdorLocalRunner that reads a task from the spec, executes the task, and
reports results.
    * Add two SPEC examples: consume-bench-spec.json and
produce-bench-spec.json.
    *  Add/update helper scripts.
    *  Update TROGDOR.md

* Run the Trogdor worker in a separate process (feature 2).
  * Users pass the worker command in spec.workerCommand, such as
"workerCommand": ["python", "./producer.py"]
  * RuntimeProcessWorker starts the worker in a process by executing
`workercommand --spec TASK_SPEC`.
  * RuntimeProcessWorker monitors the worker's stderr and stdout line by
line, if the stdout line is a JSON value, it updates its status with
the JSON value.




*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Commit:
Execute Trogdor tasks directly

It is handy to run Trogdor tasks directly without setting up
distributed Trogdor coordinator/agents. For example, we can quickly
run produce and consume performance tests.

+ Add TrogdorLocalRunner that reads a task from the spec, executes the task, and
reports results.
+ Add two SPEC examples: consume-bench-spec.json and
produce-bench-spec.json
+ Add/update helper scripts.
+ Update TROGDOR.md

hachikuji commented:
cc @cmccabe

cmccabe commented on bin/trogdor.sh:
```diff
@@ -23,6 +23,7 @@ Usage:
 
 Actions:
   agent: Run the trogdor agent.
+  local: Run the trogdor spec locality.
```
How about "exec" here?   "local" seems ambiguous, since agents also run locally.

yangxi replied:
OK

cmccabe commented on tests/trogdor/bin/trogdor-local-run.sh:
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
```
It would be good if we could add the ability to read JSON from a file to the Java code, rather than hacking it up in the shell.  The reason is because the shell has some issues with escaping special characters.  Another reason is because this would be useful more generally in all the other trogdor command line commands that take JSON arguments.

So we could just write `./bin/trogdor.sh exec --spec @/my/spec/path`

And then we could also write something like `bin/trogdor.sh client -create-task @/my/task/path`, etc.

I think argparse4j has built-in support for this:
https://argparse4j.github.io/apidocs/net/sourceforge/argparse4j/ArgumentParserBuilder.html#fromFilePrefix-java.lang.String-

yangxi replied:
Sure.

yangxi replied:
fromFilePrefix does not help here because what it does is to read arguments from a file. So you have to put "--spce {....}" into a file, then call ./bin/trogdor.sh exec @/my/spec/path.

I think it is better to read the SPEC file directly. We can also ignore lines starting with `#`, so we can put comments in the SPEC files.

yangxi replied:
@cmccabe Updated, please have a look. It reads the SPEC file if the arguments starting with `@`.

Commit:
Read Trogdor SPEC from a file.

Commit:
Update README about how to execute Trogdor SPEC tasks directly.

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchWorker.java:
```diff
@@ -360,6 +360,7 @@ StatusData update() {
                 latSummary.percentiles().get(1).value(),
                 latSummary.percentiles().get(2).value());
             statusUpdater.updateConsumeStatus(consumer.clientId(), statusData);
+            statusUpdater.update();
```
I think this call might not be needed, as we already have the `statusUpdater` scheduled to run `update()` once every minute and the `ConsumeStatusUpdater` scheduled to run once every minute.
I'm wondering if a minute is too long of a time frame and whether or not we should edit it

stanislavkozlovski replied:
Also, I know that this isn't from this PR but I just noticed it. On line 314, we have `WorkerUtils.abort(log, "ConsumeStatusUpdater", e, doneFuture);` whereas the string should be `StatusUpdater`. Is it possible we update that while we're at it?

yangxi replied:
@stanislavkozlovski  I print out the histogram at `doneFuture`. Without that `update()`, it is possible that the histogram is not updated when printing the result.

yangxi replied:
Changed the `ConsumeStatusUpdater` to `StatusUpdater`. Thanks.

Commit:
Address some warnings.

Commit:
Run the Trogdor worker in a separate process.

This patch adds the Trogdor RuntimeProcessWorker that runs the worker
in a separate process. With this, Trogdor can execute diverse
workloads, such as non-Java clients.

+ Users pass the worker command in spec.workerCommand, such as
"workerCommand": ["python", "./producer.py"]

+ RuntimeProcessWorker starts the worker in a process by executing
`workercommand --spec TASK_SPEC`.

+ RuntimeProcessWorker monitors the worker's stderr and stdout line by
line, if the stdout line is a JSON value, it updates its status with
the JSON value.

Commit:
Merge github.com:apache/kafka into local-trogdor

Commit:
Merge github.com:apache/kafka into trunk

Commit:
Merge branch 'trunk' into local-trogdor

stanislavkozlovski commented on tests/trogdor/spec/consume-bench-spec.json:
```diff
@@ -0,0 +1,28 @@
+    "targetMessagesPerSec": 100,
+    "threadsPerWorker": 5,
+    "consumerGroup": "cg",
+    "maxMessages": 100,
```
It'd be nice to increase the `maxMessages` in this example such that people that use this for tests don't have their task finish in the first second of execution.
We could maybe standardize with the `produce-bench-spec.json` to ensure both run around the same time and process around the same amount of messages?

yangxi replied:
Bump maxMessages to 500.

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/common/SpecUtils.java:
```diff
@@ -0,0 +1,58 @@
+    public static String readSpec(String argument) throws IOException {
+        if (argument.startsWith("@")) {
+            BufferedReader br = null;
+            try {
```
nit: maybe use a [try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) for instantiating and closing the `BufferedReader`?

yangxi replied:
Thanks, updated it with try-with

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/task/TaskSpec.java:
```diff
@@ -52,10 +53,26 @@
+    protected TaskSpec(
+            @JsonProperty("startMs") long startMs,
             @JsonProperty("durationMs") long durationMs) {
         this.startMs = startMs;
```
nit: Could we just call the above constructor here and avoid duplicating assignment logic? `this(startMs, durationMs, null)`

yangxi replied:
Sure, thanks.

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ConsumeBenchSpec.java:
```diff
@@ -182,7 +182,11 @@ public TaskController newController(String id) {
     @Override
     public TaskWorker newTaskWorker(String id) {
-        return new ConsumeBenchWorker(id, this);
+        if (this.workerCommand() == null) {
```
If we are to have `workerCommand()` be in `TaskSpec`, I think it would make sense to have this check and `RuntimeProcessWorker` instantiation be in `TaskSpec#newTaskWorker`. We could maybe then have a new abstract method (e.g `TaskSpec#newSpecificTaskWorker` or whatever) that inheriting classes would override

yangxi replied:
Well, the problem is that `RuntimeProcessWorker` is a special case. Starting RuntimeProcessWorker in `TaskSpec` makes it feel like that is a default worker.

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchSpec.java:
```diff
@@ -73,21 +75,24 @@
+            @JsonProperty("activeTopics") TopicsSpec activeTopics,
+            @JsonProperty("inactiveTopics") TopicsSpec inactiveTopics) {
+        super(startMs, durationMs, workerCommand);
+        System.out.println("workerComamnd is " + workerCommand);
```
nit: leftover log

yangxi replied:
Sorry, deleted.

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/workload/RuntimeProcessWorker.java:
```diff
@@ -0,0 +1,178 @@
+        this.worker = null;
+        this.status = null;
+        this.doneFuture = null;
+        doneFuture.complete("");
```
As [findBugs in the Jenkins CI complains](https://builds.apache.org/job/kafka-pr-jdk8-scala2.11/18686/findbugsResult/new/) - this will cause an NPE

yangxi replied:
My stupid code.

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/workload/RuntimeProcessWorker.java:
```diff
@@ -0,0 +1,178 @@
+                executor.submit(new ErrorUpdater());
+                worker.waitFor();
+                if (worker.exitValue() != 0) {
+                    status.update(new TextNode("Error: " + command + " returns " + worker.exitValue()));
```
nit: "returns" => "returned" ?

yangxi replied:
Yeah

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/workload/RuntimeProcessWorker.java:
```diff
@@ -0,0 +1,178 @@
+    public class ErrorUpdater implements Runnable {
+        BufferedReader br;
+        ErrorUpdater() {
+            br = new BufferedReader(new InputStreamReader(worker.getErrorStream(), StandardCharsets.UTF_8));
```
Does this get closed automatically by Java or are we leaking it? As far as I know this would keep the system resources until it's closed, but I guess garbage collection would eventually close it? Same question for `StatusUpdated#br`

yangxi replied:
Yeah, I think so since the stream is gone after either the process is gone.

stanislavkozlovski replied:
Which process? Trogdor Agent processes are long-lived

yangxi replied:
The worker process. After it is gone, its stdout and stderr streams should be closed.

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/workload/RuntimeProcessWorker.java:
```diff
@@ -0,0 +1,178 @@
+        if (!running.compareAndSet(true, false)) {
+            throw new IllegalStateException("ProduceBenchWorker is not running.");
+        }
+        log.info("{}: Deactivating ProduceBenchWorker.", id);
```
typo `ProduceBenchWorker` and in the exception above

yangxi replied:
Thanks

stanislavkozlovski commented on tools/src/main/java/org/apache/kafka/trogdor/workload/RuntimeProcessWorker.java:
```diff
@@ -0,0 +1,178 @@
+                    log.info("Worker (stdout):{}", line);
+                    try {
+                        JsonNode resp = JsonUtil.JSON_SERDE.readTree(line);
+                        status.update(resp);
```
There's just one thing I'm not sure about here. In `ConsumeBenchWorker` and `RoundTripWorker` we update the status every 30 seconds. In `ProduceBenchWorker` we update the status every minute.
Here, we update it on every single line which means that it might be very easy to miss certain outputs when monitoring this task through the coordinator.

I'm wondering whether it would make sense to update this status once every 30s/1m with a concatenation of all the read json in that time frame (up to a certain limit I imagine).}
I'd like to hear your thoughts on this @yangxi @cmccabe

yangxi replied:
Both `ConsumeBenchWorker` and `ProduceBenchWorker` overwrites the status, and both expect the worker returns final results at the end of task. This follows the same idea. It does not update status for every line, but only updates the status if the result is a JSON string. Which means that we expect the worker know this and only return the final results in JSON string at the end of the task.

Let's keep it like this. If overwriting becomes a problem, we can do some improvements:
1) We can decide the response API. For example, we only update the status for lines like this {task: TASK_ID, response: { ..... } } 
2) We can concatenate output lines.
3) We can send out status as a Kafka stream :)

stanislavkozlovski replied:
They do overwrite the status but they do it once every 30 seconds, so as a user you know that if you query once every 30 seconds you'll get the latest status.
In this case you don't, you may query once every 30 seconds but you're not sure whether some message got written and then overwritten in that timespan.

@cmccabe is working on https://issues.apache.org/jira/browse/KAFKA-7793, which are usability improvements to the Trogdor CLI and I thought that having a way to follow a tasks' status might be useful. If we settled on a pre-defined status update time we could implement that easily.

Anyway that's just my thoughts on this. In the end, we do log every line via`log.info("Worker (stdout):{}", line);` so it shoudn't hurt

stanislavkozlovski reviewed (commented):
Thanks a lot for the PR, Xi!
These changes (especially `RuntimeProcessWorker `) are very useful and will unlock a lot of use cases with Trogdor. It's looking great to me.

I left some small comments and want to discuss the status of the `RuntimeProcessWorker` a bit more

Commit:
Minor fixes

Commit:
Reuse the constructor.

cmccabe commented:
Hi @yangxi,

Thanks for a great idea for a PR, and for taking the time to put this together.

As I was working on #6133, I thought it would be easier to just roll this change into that one.  The reasoning is because they touch on a lot of the same files.  For example, that patch removes the "start a task" bash scripts in favor of just being able to specify a file with JSON in it via `trogdor.sh client startTask -i taskId -s /path/to/spec`.  I think this is a much cleaner approach which eliminates the need for a lot of bash glue code.

The approach in #6133 is a bit different because it just adds an `--exec` command-line flag to the Agent process, rather than creating a separate process.  When the agent is given this `--exec` flag, it immediately starts the given task spec, and waits for it to complete.

I think that is the right way to go, since it means that we have fewer different ways of doing the same thing.  In particular, the command-line tools for querying the agent over its REST API will continue to work with the change in #6133, whereas they wouldn't here if we were creating a separate process.  This could be very useful if you want to get status updates over time of something that you are running.

yangxi commented:
@cmccabe 
> Hi @yangxi,
> As I was working on #6133, I thought it would be easier to just roll this change into that one. The reasoning is because they touch on a lot of the same files. For example, that patch removes the "start a task" bash scripts in favor of just being able to specify a file with JSON in it via `trogdor.sh client startTask -i taskId -s /path/to/spec`. I think this is a much cleaner approach which eliminates the need for a lot of bash glue code.

This PR reads from the JSON config file too, if the spec name starts with `@`, it reads the file, ignores lines starting with `#`. 

Are you going to commit the #6133 soon, I can make the TrogdorLocalRunner use same arguments.

> 
> The approach in #6133 is a bit different because it just adds an `--exec` command-line flag to the Agent process, rather than creating a separate process. When the agent is given this `--exec` flag, it immediately starts the given task spec, and waits for it to complete.

Sorry about that this PR squeezes two not related features:  1) directly execute Trogdor tasks without setting up coordinators and agents, 2) make agents be able to execute shell commands as the worker in another process, so we have the flexibility of running workloads, such as non-Java clients, docker images, etc.

I guess you want to compare #6133 with the feature 1.  For feature 1, it does not create a separate process to run the spec. It runs the task directly.

> 
> I think that is the right way to go, since it means that we have fewer different ways of doing the same thing. In particular, the command-line tools for querying the agent over its REST API will continue to work with the change in #6133, whereas they wouldn't here if we were creating a separate process. This could be very useful if you want to get status updates over time of something that you are running.

Sorry for the confusion of combing two un-related features in one PR.
Feature 1 directly executes the tasks. The goal is to run tasks without setting up agents, etc. So there is no need to query the results via REST API since you run it locally.

Feature 2 does not change how agents work. The agent starts the worker process, and the agent updates results same as other agents. So, you just query results as same as before from agents.

stanislavkozlovski commented:
> Sorry for the confusion of combing two un-related features in one PR.
Feature 1 directly executes the tasks. The goal is to run tasks without setting up agents, etc. So there is no need to query the results via REST API since you run it locally.

It might make sense to query for the results if your task were to run for 1 hour, for example

yangxi commented:
> > Sorry for the confusion of combing two un-related features in one PR.
> > Feature 1 directly executes the tasks. The goal is to run tasks without setting up agents, etc. So there is no need to query the results via REST API since you run it locally.
> 
> It might make sense to query for the results if your task were to run for 1 hour, for example

You are right. Adding `--exec` is a better approach. Looks like we don't need the `TrogdorLocalRunner` then. I wish I knew #6133.

@cmccabe Please have a look the second feature (run the Trogdor worker in a separate process). If you think it is helpful, I can submit it as a PR to your #6133 branch.
